File: backend/app/core/dtos/google_calendar.py
from datetime import datetime

from pydantic import BaseModel
from pydantic.fields import Field

from app.core.dtos.base import BaseModelWithErrorCodes
from app.core.features.google_calendar import GoogleCalendarSyncStatus


class FolloweeCalendarInfo(BaseModel):
    username: str = Field(..., title="Followee Username")
    nickname: str | None = Field(None, title="Followee Nickname")
    calendar_url: str = Field(..., title="Public Calendar URL")


class GetGoogleCalendarAuthUrlResponse(BaseModelWithErrorCodes):
    authorization_url: str = Field(..., title="Google OAuth Authorization URL")


class HandleGoogleCalendarOAuthCallbackRequest(BaseModel):
    auth_code: str = Field(..., title="Google OAuth Authorization Code")


class HandleGoogleCalendarOAuthCallbackResponse(BaseModelWithErrorCodes):
    integration_id: str | None = Field(None, title="Integration ID")
    calendar_url: str | None = Field(None, title="Public Calendar URL")


class GetGoogleCalendarStatusResponse(BaseModelWithErrorCodes):
    integration_id: str | None = Field(None, title="Integration ID")
    google_email: str | None = Field(None, title="Google Email")
    calendar_url: str | None = Field(None, title="Public Calendar URL")
    sync_status: GoogleCalendarSyncStatus | None = Field(None, title="Sync Status")


class DisconnectGoogleCalendarResponse(BaseModelWithErrorCodes):
    pass


class SyncGoogleCalendarResponse(BaseModelWithErrorCodes):
    sync_status: GoogleCalendarSyncStatus | None = Field(None, title="Sync Status")
    events_synced: int | None = Field(None, title="Number of Events Synced")


class GetFolloweeCalendarsResponse(BaseModelWithErrorCodes):
    calendars: list[FolloweeCalendarInfo] = Field(default_factory=list, title="Followee Calendars")

File: frontend/src/lib/api/dtos/google-calendar.ts
export type GoogleCalendarSyncStatus = "disconnected" | "connected" | "syncing" | "error";

export interface FolloweeCalendarInfo {
  username: string;
  nickname: string | null;
  calendar_url: string;
}

export interface GetGoogleCalendarAuthUrlResponse {
  authorization_url: string;
  error_codes: number[];
}

export interface HandleGoogleCalendarOAuthCallbackRequest {
  auth_code: string;
}

export interface HandleGoogleCalendarOAuthCallbackResponse {
  integration_id: string | null;
  calendar_url: string | null;
  error_codes: number[];
}

export interface GetGoogleCalendarStatusResponse {
  integration_id: string | null;
  google_email: string | null;
  calendar_url: string | null;
  sync_status: GoogleCalendarSyncStatus | null;
  error_codes: number[];
}

export interface DisconnectGoogleCalendarResponse {
  error_codes: number[];
}

export interface SyncGoogleCalendarResponse {
  sync_status: GoogleCalendarSyncStatus | null;
  events_synced: number | null;
  error_codes: number[];
}

export interface GetFolloweeCalendarsResponse {
  calendars: FolloweeCalendarInfo[];
  error_codes: number[];
}
